std/math.d

```diff
@@ -5884,14 +5885,285 @@ T nextafter(T)(const T x, const T y) @safe pure nothrow @nogc
+    assert(fmin(1) == 1.0);
+    assert(fmin(ubyte.max, 25.0f) == 25.0f);
+    // Static and dynamic floating-point arrays, too:
+    double[7] nums = [0.0, double.min_normal, 0.5, -1.0, -double.max, double.infinity, 8];
+    assert(fmin(nums) == -double.max);
+    assert(fmin(nums[1 .. 3]) == double.min_normal);
+
+    // NaN handling:
+    assert(fmin().isNaN);
+    assert(fmin(13, -431_315, long.max, real.nan, float.infinity).isNaN);
+
+    // The return type is always floating-point, and mixed signed-unsigned comparisons
+    // are safe, even if all the arguments are integers.
+    assert(fmin(-1, cast(uint)int.max) is -1.0);
+    // The F template parameter can be used to request a specific floating-point type:
+    static assert(is(typeof(fmin!float(1u)) == float));
+    static assert(is(typeof(fmin!double(1L)) == double));
+    static assert(is(typeof(fmin!real(1.0f)) == real));
+}
+
+private template AutoFImumType(T ...)
+{
+    private enum int reqPrec =
+    {
+        int max = (T.length == 0) ? double.mant_dig : 0;
+        foreach (V; T)
+        {
+            int prec;
+            static if (is(V : short) || is(V : ushort))
+                prec = 8 * ushort.sizeof;
+            else static if (is(V : int) || is(V : uint))
+                prec = 8 * uint.sizeof;
+            else static if (is(V : long) || is(V : ulong))
+                prec = 8 * ulong.sizeof;
+            else static if (is(typeof(typeof(V.init + 1).mant_dig)))
+                prec = typeof(V.init + 1).mant_dig;
+            else
+                prec = real.mant_dig;
+
+            if (prec > max)
+                max = prec;
+        }
+        return max;
+    }();
+
+    static if (reqPrec <= float.mant_dig)
+        alias AutoFImumType = float;
+    else static if (reqPrec <= double.mant_dig)
+        alias AutoFImumType = double;
+    else
+        alias AutoFImumType = real;
+}
+
+unittest
+{
+    foreach (F; AliasSeq!(float, double, real))
+        static assert(is(AutoFImumType!F == F));
+    foreach (I; AliasSeq!(bool, byte, ubyte, char, short, ushort, wchar))
+        static assert(is(AutoFImumType!I == float));
+    foreach (I; AliasSeq!(int, uint, dchar))
+        static assert(is(AutoFImumType!I == double));
+    foreach (I; AliasSeq!(long, ulong))
+        static assert(is(AutoFImumType!I == real));
+
+    enum Foo : float { foo = 1.0f }
+    static assert(is(AutoFImumType!Foo == float));
+    enum Bar : long { bar = 1 }
+    static assert(is(AutoFImumType!Bar == real));
+
+    struct Baz
+    {
+        short baz;
+        alias baz this;
+    }
+    static assert(is(AutoFImumType!Baz == float));
+
+    static assert(is(AutoFImumType!(ushort, Foo) == float));
+    static assert(is(AutoFImumType!(uint, wchar) == double));
+    static assert(is(AutoFImumType!(Baz, uint, float) == double));
+    static assert(is(AutoFImumType!(byte, Bar, long) == real));
+    static assert(is(AutoFImumType!(real, int) == real));
+}
```

Comment:
There are an infinite number of possible combinations, because `T` can be any length. What's special about those combinations that needs to be tested?